dialpolicy.go

```diff
@@ -0,0 +1,79 @@
+package autonat
+
+import (
+	"github.com/libp2p/go-libp2p-core/host"
+	ma "github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr-net"
+)
+
+type dialPolicy struct {
+	allowSelfDials bool
+	host           host.Host
+}
+
+// skipDial indicates that a multiaddress isn't worth attempted dialing.
+// The same logic is used when the autonat client is considering if
+// a remote peer is worth using as a server, and when the server is
+// considering if a requested client is worth dialing back.
+func (d *dialPolicy) skipDial(addr ma.Multiaddr) bool {
+	// skip relay addresses
+	_, err := addr.ValueForProtocol(ma.P_CIRCUIT)
+	if err == nil {
+		return true
+	}
+
+	if d.allowSelfDials {
+		return false
+	}
+
+	// skip private network (unroutable) addresses
+	if !manet.IsPublicAddr(addr) {
+		return true
+	}
+	candidateIP, err := manet.ToIP(addr)
+	if err != nil {
+		return true
+	}
+
+	// Skip dialing addresses we believe are the local node's
+	for _, localAddr := range d.host.Addrs() {
```

Comment:
note: let's avoid calling `host.Addrs()` repeatedly.

Reply:
I'd actually just change these methods to free functions and pass in our current addrs.

Reply:
it's useful for tests to have a test flag for allowing self dials. it seems weird to pass a testing boolean flag into the functions on every invocation. ideas for that?

Reply:
Ah, good point. I'm still a bit concerned about calling this once per peer address, but we can punt that till later.